### PR TITLE
Update default engine to FAISS

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
- Update default engine to FAISS.
- Fixed tests that assumed default engine to be nmslib to either use nmslib explicitly or update test assertion to look for faiss specific artifacts like file extension instead of nmslib artifacts
- removed tests that are just validating default (provided it is nmslib) doesn't support certain features
### Related Issues
#2163

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
Doc issue: https://github.com/opensearch-project/documentation-website/issues/8589

**Review by vibrantvarun — COMMENTED** [review]
Looks good! Will wait for other engineers review

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java:206]: Keep as NMSLIB - nmslib is only legacy support

**Comment by jmazanec15** [comment]
Can you update this too: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/engine/EngineResolver.java#L51

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java:206]: maybe add comment to like I should have done haha

File: src/main/java/org/opensearch/knn/index/engine/KNNEngine.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/engine/KNNEngine.java:32]: How we are handling the backward compatibility here? Also can we validate if we have proper BWC tests to test this default engine change.

File: src/test/java/org/opensearch/knn/index/KNNIndexShardTests.java
Comment (shatejas) [inline src/test/java/org/opensearch/knn/index/KNNIndexShardTests.java:118]: is the compound extension covered?

File: src/test/java/org/opensearch/knn/index/VectorDataTypeIT.java
Comment (shatejas) [inline src/test/java/org/opensearch/knn/index/VectorDataTypeIT.java:267]: Just checking if this case is covered when engine is explicitly NMSLIB

File: src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java
Comment (shatejas) [inline src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:149]: Out of curiosity, why this change?

File: src/main/java/org/opensearch/knn/index/engine/KNNEngine.java
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/engine/KNNEngine.java:32]: I was just curious to know , why do we need to handle BWC in this case , So i created a Index , without specifying any engine so it picked NMSLIB , so now this information is already there in Cluster State and Serialised and passed to all Nodes, Now During query time , we do get it information from there and it does not depend on the new Default.

File: src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java
Comment (jmazanec15) [reply src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:149]: @vamshin is it okay default engine doesnt support cosine?

Comment (jmazanec15) [reply src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:149]: We could add this by taking a normalization and then setting to innerproduct. But this would take some effort.

Comment (vamshin) [reply src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:149]: ah good catch! I think still we need to make faiss as default based on all the advanced capabilities like nested fields, efficient filters, disk optimized search, in future GPUs. In long term makes sense to build the normalization.

File: src/main/java/org/opensearch/knn/index/engine/KNNEngine.java
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/engine/KNNEngine.java:32]: >  So i created a Index , without specifying any engine so it picked NMSLIB , so now this information is already there in Cluster State and Serialised and passed to all Nodes,

the engine gets persisted with indexMapping only when you create the index and also provide the key `method`. But if you don't specify the key `method` while creating the index and lets say create an index like this:
```
PUT my-knn-index-1
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
        "my_vector1": {
          "type": "knn_vector",
          "dimension": 2
        }
    }
  }
}
```

then we infer the engine. Ref: https://github.com/opensearch-project/k-NN/blob/HEAD/src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperUtil.java#L196-L219 for more details.

File: src/test/java/org/opensearch/knn/index/VectorDataTypeIT.java
Comment (VijayanB) [reply src/test/java/org/opensearch/knn/index/VectorDataTypeIT.java:267]: Yes, it is covered explicitly for nmslib engine

File: src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java
Comment (VijayanB) [reply src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:149]: @shatejas FAISS doesn't support cosine yet.

File: src/test/java/org/opensearch/knn/index/KNNIndexShardTests.java
Comment (VijayanB) [reply src/test/java/org/opensearch/knn/index/KNNIndexShardTests.java]: Updated test to use nmslib

**Review by jmazanec15 — APPROVED** [review]

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java:552]: why this is static? and could we move this function to : KNNVectorFieldMapperUtil class?

**Review by navneet1v — APPROVED** [review]
I just have a minor comment. I see

**Review by vibrantvarun — APPROVED** [review]

File: src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java
Comment (VijayanB) [reply src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java]: Ack. Moved to Util class

**Review by jmazanec15 — DISMISSED** [review]

**Review by navneet1v — APPROVED** [review]

**Comment by VijayanB** [comment]
Tests are failing not related to this PR. This PR https://github.com/opensearch-project/k-NN/pull/2236 will fix those issue in main. Will rebase once the other PR is merged.

**Comment by VijayanB** [comment]
Test failure is not due to this change.

**Review by jmazanec15 — APPROVED** [review]

**Review by naveentatikonda — APPROVED** [review]

**Comment by VijayanB** [comment]
Failure is due to waiting for this PR to be backported https://github.com/opensearch-project/k-NN/pull/2248

**Pull request merged**
